perceval/backends/stackexchange.py

```diff
@@ -0,0 +1,296 @@
+
+        cache_items = self.cache.retrieve()
+
+        for question in cache_items:
+            yield question
+
+
+class StackExchangeClient:
+    """StackExchange API client.
+
+    This class implements a simple client to retrieve questions from
+    any Stackexchange site.
+
+    :param site: URL of the Bugzilla server
+    :param tagged: filter items by question Tag
+    :param token: StackExchange access_token for the API
+    :param max_questions: max number of questions per query
+
+    :raises HTTPError: when an error occurs doing the request
+    """
+
+    def __init__(self, site, tagged, token, max_questions):
+        self.site = site
+        self.tagged = tagged
+        self.token = token
+        self.max_questions = max_questions
+        self.version = VERSION_API
+
+    def __build_base_url(self, type='questions'):
+        base_api_url = STACKEXCHANGE_API_URL
+        base_api_url = urljoin(base_api_url, self.version, type)
+        return base_api_url
+
+    def __build_payload(self, page, from_date, order='desc', sort='activity'):
+        payload = {'page': page,
+                   'pagesize': self.max_questions,
+                   'order': order,
+                   'sort': sort,
+                   'tagged': self.tagged,
+                   'site': self.site,
+                   'key': self.token,
+                   'filter': QUESTIONS_FILTER}
+        if from_date:
+            timestamp = int(time.mktime(from_date.timetuple()))
+            payload['min'] = timestamp
+        return payload
+
+    def __log_status(self, quota_remaining, quota_max, page_size, total):
+
+        logger.info("Rate limit: %s/%s" % (quota_remaining,
+                                           quota_max))
+        if (total != 0):
+            if (total <= page_size):
+                logger.info("Fetching questions: %s/%s" % (total,
+                                                           total))
+            else:
+                logger.info("Fetching questions: %s/%s" % (page_size,
+                                                           total))
+        else:
+            logger.info("No questions were found.")
+
+    def get_questions(self, from_date):
+        """Retrieve all the questions from a given date.
+
+        :param from_date: obtain questions updated since this date
+        """
+
+        page = 1
+        req = requests.get(self.__build_base_url(),
+                           params=self.__build_payload(page, from_date))
+        req.raise_for_status()
+        questions = req.json()['items']
+
+        if (req.json()['page_size'] >= req.json()['total']):
+            fetched = req.json()['total']
+        else:
+            fetched = req.json()['page_size']
+
+        self.__log_status(req.json()['quota_remaining'],
+                          req.json()['quota_max'],
+                          fetched,
+                          req.json()['total'])
+
+        while questions:
```

Comment:
This whole loop has to be outside, in the backen class. This is not the duty of the client, it's the duty of the backend to know when it has to asks for more items. The rationale behind this it's we want to have clients that can be used to ask for anything (for instance, to ask for the third page of a certain set of questions without having to start from the first page). This make the code reusable by others and easier to maintain and to fix.